### ISSUE TITLE
RfC: Add GitHub Actions Workflow to Publish Briefcase to PyPI

### ISSUE THREAD

**Issue by danyeaw** [beeware/briefcase#281]
As part of #266, @freakboy3742 swapped to using GitHub Actions for CI. One thing that we were chatting about that would be nice is to add a CD component in order to publish a new version to PyPI.

This issue to collect feedback on this update. I think there are four main components to this discussion:

1. What events we want it to push to PyPI on?
2. Do we want to use setuptools, or one of the more recent tools like Twine or Poetry?
3. Do we want to upload any artifacts to the Release on GitHub?
4. Do we want to do some checks after it is uploaded like try to pip install the library on all the platforms?

For 1: I think we have these main options:
1. Any time a new tag in the format of `v#.#.#` is pushed to the repo
2. Any time a new Release is created, this is the default that the Publish Python Package GitHub Actions template uses
3. Any time a new Release is published
This library isn't currently making use of adding release notes or uploading assets to the releases. I think going with the 2nd option would be a straightforward approach in this case. You would push your tag, draft a release which kicks off the workflow, and then once the workflow is done manually publish the release.

For 2:
I see the options are leave things alone and use setuptools, go all-in and use Poetry (with `poetry config virtualenvs.create false` or `poetry config virtualenvs.in-project = true` so that it doesn't take over virtualenv creation), or use Twine just for building and uploading to PyPI.

For 3: since artifacts aren't currently part of the Releases, I would assume we don't start now.

For 4: this is more of a nice to have, but could give us more confidence in a release after it is pushed to PyPI.

**Comment by freakboy3742** [comment]
Thanks for putting this together, @danyeaw. 

There's a part of me that likes the simplicity of (1) and cutting a release by just pushing a tag. However, given that it's almost impossible to undo mistakes made on PyPI, I'd feel better if there was a slightly more deliberate, human-reviewable action in place. 

I tend to agree that (2) is the best of the options you present. However, an even better option might be a hybrid: create a *draft* release whenever vX.X.X is tagged, but formally publish to PyPI when the release is published. That automates the process setting up and formally publishing the release, but makes the *trigger* of the actual publication a manual, human-approved process. If adopting a tool like [towncrier](https://github.com/hawkowl/towncrier) for release notes will help automating the drafting process, I'd be open to doing that.

I'm not sure what you're driving at with your discussion of setuptools, twine and poetry; as far as I'm aware, those are three independent tools, solving different problems:
 * setuptools defines the packaging details for the library. We'll definitely need to use setuptools (since there's a setup.py, and we need to create a wheel)
 * twine provides the tools to upload the package to PyPI. Again, we're definitely going to need to use this. The only option for *not* using twine would be to use `setuptools upload`, but that has been specifically deprecated in favour of twine.
 * poetry lets you lock down the dependencies in a runtime environment. We're defining a library to be installed, so there shouldn't be a lockfile. Poetry can't be used to build wheels or upload them. best case, it would be providing the environment in which wheels can be constructed, and I'm still waiting for the poetry/pipenv/pip.lock dust to settle some more (and for some of the UX rough edges to be worn down) before I commit to one of these tools.

Uploading artefacts to the release isn't a huge priority for me personally. Given that PyPI is the canonical source for Python artefacts, I'd rather not muddy the waters by having multiple artefact repositories unless there's a particularly compelling reason to do so.

I agree post-release checks would be a nice-to-have, at best. If that step reveals any problems, then something has gone *badly* wrong. However, it might be nice to know about those problems as soon as they happen, rather than when a user encounters them. 

Looking at Briefcase specifically, I want to add a CI test that does a create/build on an empty hello-world project (i.e., essentially doing the first steps of the tutorial). Running something like that check after a release on a version of the library obtained from PyPI could be a useful check.

**Comment by danyeaw** [comment]
@freakboy3742 I like that, I will set it up to draft the release when a tag is pushed, and then upload to PyPI when the release is published.

Poetry is a dependency management tool _and_ a tool to interact with PyPI. You can run `poetry build` and `poetry publish` and you have your new library in PyPI, pretty similar to Twine. I understand you want to stick with setuptools and Twine, and that should work good. I also wasn't aware that `setuptools upload` was deprecated, so thanks for updated me on that.

I'll hold off on the post-release checks for now, and we can add that as a second step if we decide we need it.

**Comment by freakboy3742** [comment]
Resolved by #282.
